src/shogun2-core/src/main/java/de/terrestris/shogun2/model/tree/TreeNode.java

```diff
@@ -0,0 +1,381 @@
+/**
+ *
+ */
+package de.terrestris.shogun2.model.tree;
+
+import javax.persistence.Entity;
+import javax.persistence.Inheritance;
+import javax.persistence.InheritanceType;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.builder.ToStringStyle;
+
+import com.fasterxml.jackson.annotation.JsonIdentityInfo;
+import com.fasterxml.jackson.annotation.JsonIdentityReference;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.ObjectIdGenerators;
+
+import de.terrestris.shogun2.converter.TreeFolderIdResolver;
+import de.terrestris.shogun2.model.PersistentObject;
+
+/**
+ * A class representing a node in a tree. This class can be used for leaf nodes.
+ *
+ * For folders, {@link TreeFolder} should be used.
+ *
+ * @author Nils Bühner
+ * @author Kai Volland
+ * @author terrestris GmbH & Co. KG
+ *
+ */
+@Entity
+@Table
+@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
+@JsonInclude(Include.NON_NULL)
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
+public class TreeNode extends PersistentObject {
+
+	/**
+	 *
+	 */
+	private static final long serialVersionUID = 1L;
+
+	/**
+	 * The text to show on node label (html tags are accepted)
+	 */
+	private String text;
+
+	/**
+	 * This is the owning side of the relation between parent/child!
+	 */
+	@ManyToOne
+	@JoinColumn(name="PARENTFOLDER_ID")
+	@JsonIdentityInfo(
+		generator = ObjectIdGenerators.PropertyGenerator.class,
+		property = "id",
+		resolver = TreeFolderIdResolver.class
+	)
+	@JsonIdentityReference(alwaysAsId = true)
+	@JsonProperty("parentId")
+	private TreeFolder parentFolder;
+
+	/**
+	 * The position of the node inside its parent. When parent has 4 children
+	 * and the node is third amongst them, index will be 2 -> starting with
+	 * index 0
+	 */
+	private int index;
+
+	/**
+	 * True if this is the root node.
+	 */
+	private boolean root;
+
+	/**
+	 * Set to true to indicate that this child can have no children. The expand
+	 * icon/arrow will then not be rendered for this node.
+	 *
+	 * As this class is used for leafs, we'll set the leaf property to true by default.
+	 * {@link TreeFolder}s will set this property to false in the constructor.
+	 */
+	private boolean leaf = true;
+
+	/**
+	 * Control checkboxes:
+	 * 
+	 * <ul>
+	 *   <li>null: no checkbox will appear</li>
+	 *   <li>true: a checked checkbox will appear</li>
+	 *   <li>false: an unchecked checkbox will appear</li>
+	 * </ul>
+	 */
+	private Boolean checked;
```

Comment:
The ternary differentiation also relies on the fact that EXT handles it this way, i.e. we can easily set a value in our model and EXT will behave as documented/expected.

We could introduce an `expandable` property, but this would require some extra lines of code for a successful interaction with EXT.